File: spa-sdk/packages/spa-sdk/src/initialize.ts
/*
 * Copyright 2019 Hippo B.V. (http://www.onehippo.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Configuration } from './api';
import { ComponentFactory, ContentFactory, Content, PageModel, Page } from './page';
import { PageModelUrlBuilder } from './url';

/**
 * Initializes the page model.
 *
 * @param modelUrlBuilder Function generating an API URL based on the current request.
 * @param componentFactory Factory to produce component entities.
 * @param contentFactory Factory to produce content items.
 * @param config Configuration of the SPA integration with brXM.
 */
export async function initialize(
  modelUrlBuilder: PageModelUrlBuilder,
  componentFactory: ComponentFactory,
  contentFactory: ContentFactory,
  config: Configuration,
): Promise<Page> {
  const url = modelUrlBuilder(config.request, config.options);
  const model: PageModel = await config.httpClient({
    url,
    method: 'get',
    headers: config.request.headers,
  });
  const root = componentFactory.create(model.page);
  const content = new Map<string, Content>();

  if (model.content) {
    Object.entries(model.content)
      .forEach(([alias, model]) => content.set(alias, contentFactory.create(model)));
  }

  return new Page(model, root, content);
}

File: repository/test/src/main/java/org/hippoecm/repository/sample/Initialize.java
/*
 *  Copyright 2008-2013 Hippo B.V. (http://www.onehippo.com)
 * 
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 * 
 *       http://www.apache.org/licenses/LICENSE-2.0
 * 
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.hippoecm.repository.sample;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.jcr.RepositoryException;
import javax.jcr.Session;
import javax.jcr.Node;
import org.hippoecm.repository.api.HippoNodeType;
import org.hippoecm.repository.HippoRepository;
import org.hippoecm.repository.HippoRepositoryFactory;

public class Initialize
{

    protected final Logger log = LoggerFactory.getLogger(Initialize.class);

    HippoRepository repository;

    private Initialize() throws RepositoryException {
        repository = HippoRepositoryFactory.getHippoRepository();
    }

    private Initialize(String location) throws RepositoryException {
        repository = HippoRepositoryFactory.getHippoRepository(location);
    }

    private void initializeRepository() throws RepositoryException {
        Session session = repository.login("dummy", "dummy".toCharArray());
        Node node = session.getRootNode();
        if(!node.hasNode("configuration"))
            node = node.addNode(HippoNodeType.CONFIGURATION_PATH,HippoNodeType.NT_CONFIGURATION);
        else
            node = node.getNode(HippoNodeType.CONFIGURATION_PATH);
        if(!node.hasNode("initialize"))
            node = node.addNode(HippoNodeType.INITIALIZE_PATH,HippoNodeType.NT_INITIALIZEFOLDER);
        else
            node = node.getNode(HippoNodeType.INITIALIZE_PATH);
        node = node.addNode("newsmodel",HippoNodeType.NT_INITIALIZEITEM);
        node.setProperty(HippoNodeType.HIPPO_NODETYPESRESOURCE,"newsmodel.cnd");
        node.setProperty(HippoNodeType.HIPPO_CONTENTRESOURCE,"navigation.xml");
        session.logout();
    }

    public static void main(String[] args) {
        Initialize bootstrap = null;
        try {
            String location = args.length>0 ? args[0] : "rmi://localhost:1099/hipporepository";
            if(location != null)
                bootstrap = new Initialize(location);
            else
                bootstrap = new Initialize();
            bootstrap.initializeRepository();
        } catch(RepositoryException ex) {
            System.err.println(ex.getClass().getName()+": "+ex.getMessage());
            ex.printStackTrace(System.err);
        }
    }
}
